### PR TITLE
added support for deriving in a pragma

### PR DESCRIPTION
**Pull request by jmchapman**
I followed @UlfNorell's suggestion https://github.com/agda/agda2hs/issues/42#issuecomment-742502084

Eg.,
```
data Exp (v : Set) : Set where
  Plus : Exp v → Exp v → Exp v
  Lit : Nat → Exp v
  Var : v → Exp v
{-# COMPILE AGDA2HS Exp deriving (Show,Eq) #-}
```
->
```
data Exp v = Plus (Exp v) (Exp v)
           | Lit Natural
           | Var v
               deriving (Show, Eq)
```

Fixes https://github.com/agda/agda2hs/issues/42

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -337,22 +338,33 @@ processPragma qn = getUniqueCompilerPragma pragmaName qn >>= \case
...
     return $ ClassPragma YesCode
-  _                                        -> return DefaultPragma
-
+  Just (CompilerPragma _ s) | "deriving" `isPrefixOf` s ->
+    -- parse a deriving clause for a datatype by tacking it onto a
+    -- dummy datatype and then only keeping the deriving part
+    case Hs.parseDecl ("data X = X " ++ s) of
+      Hs.ParseFailed loc msg ->
+        setCurrentRange (srcLocToRange loc) $ genericError msg
+      Hs.ParseOk m | Hs.DataDecl _ _ _ _ _ ds <- m ->
```
Comment (UlfNorell) [inline Main.hs:347]:
```suggestion
      Hs.ParseOk (Hs.DataDecl _ _ _ _ _ ds) ->
```

File: Main.hs [hunk 2]
```diff
@@ -337,22 +338,33 @@ processPragma qn = getUniqueCompilerPragma pragmaName qn >>= \case
...
+        return $ DerivingPragma ds
+      Hs.ParseOk _ -> __IMPOSSIBLE__
+  _ -> return DefaultPragma
 compile :: Options -> ModuleEnv -> IsMain -> Definition -> TCM CompiledDef
 compile _ _ _ def = processPragma (defName def) >>= \ case
   NoPragma -> return []
   DefaultPragma -> compile' def
   ClassPragma NoCode -> return []
   ClassPragma YesCode -> return []
+  DerivingPragma ds   -> tag <$> compileData (map (() <$) ds) def
```
Comment (UlfNorell) [inline Main.hs:357]: We haven't checked that the def is a `Datatype` here, so will hit the `__IMPOSSIBLE__` in `compileData` if you put `deriving` on a non-datatype. I wonder if a nicer solution would be to pass the pragma to `compile'`. That would also get rid of the duplication of `tag`.

**Review by UlfNorell — APPROVED** [review]

**Comment by UlfNorell** [comment]
Don't forget to put a "fixes #42" in the description so it auto closes the issue when merged.

File: Main.hs [hunk 3]
```diff
@@ -337,22 +338,33 @@ processPragma qn = getUniqueCompilerPragma pragmaName qn >>= \case
...
+      where tag code = [(nameBindingSite $ qnameName $ defName def, code)]
```
Comment (omelkonian) [inline Main.hs:358]: I think it makes sense to merge `compile` with `compile'`, I don't see any use-sites that uses only one of them.

File: Main.hs [hunk 4]
```diff
@@ -337,22 +338,33 @@ processPragma qn = getUniqueCompilerPragma pragmaName qn >>= \case
     return $ ClassPragma NoCode
   Just (CompilerPragma _ s) | s == "class" ->
     return $ ClassPragma YesCode
-  _                                        -> return DefaultPragma
-
+  Just (CompilerPragma _ s) | "deriving" `isPrefixOf` s ->
+    -- parse a deriving clause for a datatype by tacking it onto a
+    -- dummy datatype and then only keeping the deriving part
+    case Hs.parseDecl ("data X = X " ++ s) of
```
Comment (omelkonian) [inline Main.hs:344]: That's quite a hack, but it seems src-exts does not provide its internal Happy parsers readily, so I guess this is fine.

**Review by omelkonian — APPROVED** [review]
Nice!

File: Main.hs (hunk 2 above)
Comment (jmchapman) [reply Main.hs]: Thanks for spotting that! I have merged `compile` and `compile'`.

File: Main.hs (hunk 3 above)
Comment (jmchapman) [reply Main.hs]: I have merged `compile` and `compile'`.

File: Main.hs (hunk 4 above)
Comment (jmchapman) [reply Main.hs:348]: Yes :) It seemed the easiest way to do it as `Language.Haskell.Exts.Parser` parser only seems to have functions/instances for bigger lumps of AST.

**Pull request merged**